src/typechecker.erl

```diff
@@ -2206,18 +2206,49 @@ type_check_call_ty(_Env, {type_error, _}, _Args, {Name, _P, FunTy}) ->
     throw(type_error(Name, FunTy, type('fun'))).
 
 -spec type_check_call_ty_intersect(env(), _, _, _) -> {type(), env(), constraints:t()}.
-type_check_call_ty_intersect(Env, [], Args, {Name, P, FunTy}) ->
-    throw(type_error(call_intersect, P, Name, FunTy, infer_arg_types(Args, Env)));
-type_check_call_ty_intersect(Env, [Ty | Tys], Args, E) ->
-    try
-        type_check_call_ty(Env, Ty, Args, E)
-    catch
-        Error when element(1,Error) == type_error ->
-            type_check_call_ty_intersect(Env, Tys, Args, E)
+type_check_call_ty_intersect(Env, ClauseTys, Args, E = {Name, P, FunTy}) ->
+    check_call_arity(hd(ClauseTys), Args, E),
+    MatchingClausesCss = lists:foldr(fun ({fun_ty, ClauseArgTys, _, _} = Clause, Acc) ->
+                                             case args_match_clause(Args, ClauseArgTys, Env) of
+                                                 {true, Cs} ->
+                                                     [{Clause, Cs} | Acc];
+                                                 false ->
+                                                     Acc
+                                             end
+                                     end, [], ClauseTys),
+    io:format("matching:\n~p\n", [MatchingClausesCss]),
+    case MatchingClausesCss of
+        [_|_] ->
```

Comment:
Without overlapping clauses, we never have more than one match here, right?

Reply:
Likely. I'll experiment with that and see.

Reply:
So getting back to this question - we might, because of the reasons [here](https://github.com/josefs/Gradualizer/pull/514#issuecomment-1425484030).